### PR TITLE
Add command for adding secret authentication for security realm.

### PR DESCRIPTION
**Pull request by istraka**

### CODE REVIEW

File: commands/src/main/java/org/wildfly/extras/creaper/commands/security/realms/AddSecretAuthentication.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
+package org.wildfly.extras.creaper.commands.security.realms;
+
+
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/security/realms/AddSecretAuthentication.java]: Oh come on :-)

File: commands/src/main/java/org/wildfly/extras/creaper/commands/security/realms/AddSecretAuthentication.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+import org.wildfly.extras.creaper.core.online.operations.Batch;
+import org.wildfly.extras.creaper.core.online.operations.Operations;
+import org.wildfly.extras.creaper.core.online.operations.Values;
+import org.wildfly.extras.creaper.core.online.operations.admin.Administration;
+
+import java.io.IOException;
+import java.util.Base64;
+
+public final class AddSecretAuthentication extends AbstractAddSecurityRealmSubElement {
+    private String password;
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/security/realms/AddSecretAuthentication.java]: `final`?

File: commands/src/main/java/org/wildfly/extras/creaper/commands/security/realms/AddSecretAuthentication.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+        Operations ops = new Operations(ctx.client);
+
+        if (replaceExisting) {
+            Batch batch = new Batch();
+            boolean secretServerIdentityExists = ops.exists(secretServerIdentitiesAddress);
+            if (secretServerIdentityExists) {
+                batch.remove(secretServerIdentitiesAddress);
+            }
+            ops.batch(batch);
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/security/realms/AddSecretAuthentication.java]: Why the batch here? It only contains at most one operation.

File: commands/src/main/java/org/wildfly/extras/creaper/commands/security/realms/AddSecretAuthentication.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+            new Administration(ctx.client).reloadIfRequired();
+        }
+
+        Batch batch = new Batch();
+        batch.add(secretServerIdentitiesAddress, Values.empty()
+                .and("value", password));
+        ops.batch(batch);
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/security/realms/AddSecretAuthentication.java]: Again, why the batch here?

File: commands/src/main/java/org/wildfly/extras/creaper/commands/security/realms/AddSecretAuthentication.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+import org.wildfly.extras.creaper.core.offline.OfflineCommandContext;
+import org.wildfly.extras.creaper.core.online.OnlineCommandContext;
+import org.wildfly.extras.creaper.core.online.operations.Address;
+import org.wildfly.extras.creaper.core.online.operations.Batch;
+import org.wildfly.extras.creaper.core.online.operations.Operations;
+import org.wildfly.extras.creaper.core.online.operations.Values;
+import org.wildfly.extras.creaper.core.online.operations.admin.Administration;
+
+import java.io.IOException;
+import java.util.Base64;
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/security/realms/AddSecretAuthentication.java]: This is only available on Java 8 and above, no?

File: commands/src/main/java/org/wildfly/extras/creaper/commands/security/realms/AddSecretAuthentication.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+         */
+        public Builder password(String password) throws IOException {
+            this.password = new String(Base64.getEncoder().encode(password.getBytes("UTF-8")), "UTF-8");
+            return this;
+        }
+
+        /**
+         * Expected password encoded in Base64
+         */
+        public Builder secretPassword(String password) throws IOException {
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/security/realms/AddSecretAuthentication.java]: It's not really "secret". Something like `passwordBase64` would be better.

File: testsuite/standalone/src/test/java/org/wildfly/extras/creaper/commands/security/realms/AddSecretAuthenticationOnlineTest.java [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+        client.apply(addSecurityRealm);
+        assertTrue("The security realm should be created", ops.exists(TEST_REALM_ADDRESS));
+    }
+
+    @Test
+    public void add() throws Exception {
+        client.apply(new AddSecretAuthentication.Builder(TEST_REALM_NAME)
+                .password(PASSWORD)
+                .build());
+        ops.exists(TEST_SECRET_IDENTITY_ADDRESS);
```
Comment (Ladicek) [inline testsuite/standalone/src/test/java/org/wildfly/extras/creaper/commands/security/realms/AddSecretAuthenticationOnlineTest.java]: Result not used?

**Comment by Ladicek** [comment]
Now thinking about the Base64 class -- we already depend on Guava and Guava has an implementation. So please use that.

**Comment by istraka** [comment]
Thank you for objections. Issues are fixed.

File: commands/src/main/java/org/wildfly/extras/creaper/commands/security/realms/AddSecretAuthentication.java [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    public static final class Builder extends AbstractAddSecurityRealmSubElement.Builder<Builder> {
+
+        private String password;
+
+        public Builder(String securityRealmName) {
+            super(securityRealmName);
+        }
+
+        /**
+         * {@link BaseEncoding#base64()} will be used.
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/security/realms/AddSecretAuthentication.java]: How about saying `This method will encode the provided {@code password} using Base64.` ?

File: commands/src/main/java/org/wildfly/extras/creaper/commands/security/realms/AddSecretAuthentication.java [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+         */
+        public Builder password(String password) throws IOException {
+            this.password = BaseEncoding.base64().encode(password.getBytes("UTF-8"));
+            return this;
+        }
+
+        /**
+         * Expected password encoded in Base64
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/security/realms/AddSecretAuthentication.java]: How about saying `This method expects the provided {@code password} to be already encoded using Base64.` ?

File: testsuite/standalone/src/test/java/org/wildfly/extras/creaper/commands/security/realms/AddSecretAuthenticationOnlineTest.java [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+                .build());
+        fail("Secret authentication method has been already configured. Should have failed.");
+    }
+
+    @Test
+    public void replaceExisting() throws Exception {
+        client.apply(new AddSecretAuthentication.Builder(TEST_REALM_NAME)
+                .password("secret")
+                .build());
+        ops.exists(TEST_SECRET_IDENTITY_ADDRESS);
```
Comment (Ladicek) [inline testsuite/standalone/src/test/java/org/wildfly/extras/creaper/commands/security/realms/AddSecretAuthenticationOnlineTest.java]: Result not used.

File: testsuite/standalone/src/test/java/org/wildfly/extras/creaper/commands/security/realms/AddSecretAuthenticationOfflineTest.java [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+
+import java.io.File;
+
+import static org.junit.Assert.fail;
+import static org.wildfly.extras.creaper.XmlAssert.assertXmlIdentical;
+
+public class AddSecretAuthenticationOfflineTest {
+
+
+    private static final String REALM_EMPTY = ""
```
Comment (Ladicek) [inline testsuite/standalone/src/test/java/org/wildfly/extras/creaper/commands/security/realms/AddSecretAuthenticationOfflineTest.java]: Why the whitespace?

File: testsuite/standalone/src/test/java/org/wildfly/extras/creaper/commands/security/realms/AddSecretAuthenticationOnlineTest.java [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+import java.util.concurrent.TimeoutException;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+@RunWith(Arquillian.class)
+public class AddSecretAuthenticationOnlineTest {
+
+
```
Comment (Ladicek) [inline testsuite/standalone/src/test/java/org/wildfly/extras/creaper/commands/security/realms/AddSecretAuthenticationOnlineTest.java]: Why the whitespace?

**Comment by Ladicek** [comment]
OK, looks good. Please address the remaining comments and squash.

**Comment by istraka** [comment]
Done.

File: testsuite/standalone/src/test/java/org/wildfly/extras/creaper/commands/security/realms/AddSecretAuthenticationOnlineTest.java [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+    @Test
+    public void add() throws Exception {
+        client.apply(new AddSecretAuthentication.Builder(TEST_REALM_NAME)
+                .password(PASSWORD)
+                .build());
+        assertTrue("Server identity should be created", ops.exists(TEST_SECRET_IDENTITY_ADDRESS));
+        ModelNodeResult result = ops.readAttribute(TEST_SECRET_IDENTITY_ADDRESS, "value");
+        result.assertSuccess();
+        assertEquals("Password should be encoded properly.",
+                PASSWORD, new String(BaseEncoding.base64().decode(result.stringValue())));
```
Comment (Ladicek) [inline testsuite/standalone/src/test/java/org/wildfly/extras/creaper/commands/security/realms/AddSecretAuthenticationOnlineTest.java]: Never ever do `new String(byte[])`, that uses the default charset. Please use `new String(byte[], String|Charset)`.

File: testsuite/standalone/src/test/java/org/wildfly/extras/creaper/commands/security/realms/AddSecretAuthenticationOnlineTest.java [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+
+        client.apply(new AddSecretAuthentication.Builder(TEST_REALM_NAME)
+                .password(PASSWORD)
+                .replaceExisting()
+                .build());
+        assertTrue("Server identity should be created", ops.exists(TEST_SECRET_IDENTITY_ADDRESS));
+        ModelNodeResult result = ops.readAttribute(TEST_SECRET_IDENTITY_ADDRESS, "value");
+        result.assertSuccess();
+        assertEquals("Password should be encoded properly.",
+                PASSWORD, new String(BaseEncoding.base64().decode(result.stringValue())));
```
Comment (Ladicek) [inline testsuite/standalone/src/test/java/org/wildfly/extras/creaper/commands/security/realms/AddSecretAuthenticationOnlineTest.java]: Never ever do `new String(byte[])`, that uses the default charset. Please use `new String(byte[], String|Charset)`.

**Comment by Ladicek** [comment]
Sorry for not noticing it earlier.

File: commands/src/main/java/org/wildfly/extras/creaper/commands/security/realms/AddSecretAuthentication.java [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+
+        public Builder(String securityRealmName) {
+            super(securityRealmName);
+        }
+
+        /**
+         * This method will encode the provided {@code password} using Base64
+         */
+        public Builder password(String password) throws IOException {
+            this.password = BaseEncoding.base64().encode(password.getBytes("UTF-8"));
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/security/realms/AddSecretAuthentication.java]: One more thing: I think it should be documented that we're using UTF-8 here. Something like: `For the purpose of the Base64 encoding, the {@code password} is treated as a sequence of bytes in the UTF-8 encoding.`

Later, we could add an overload where one can specify the encoding themselves, but we don't need that now.

File: testsuite/standalone/src/test/java/org/wildfly/extras/creaper/commands/security/realms/AddSecretAuthenticationOnlineTest.java (hunk 14 above)
Comment (istraka) [inline testsuite/standalone/src/test/java/org/wildfly/extras/creaper/commands/security/realms/AddSecretAuthenticationOnlineTest.java]: Thanks. I forgot to do this in tests after checkstyle plugin had noticed me.

**Comment by istraka** [comment]
Done.

**Comment by Ladicek** [comment]
OK, merged, thanks!

**Pull request merged**